### DISCUSSION TITLE
Kafka schema guessing timeout error

### DISCUSSION THREAD

**Discussion by mihau1211** [apache/streampipes#2280] [Q&A] [answered]
Hi guys, I have an issue reading data from the Kafka topic.

The thing is that I have Kafka connected to the Kafka Adapter, all topics are listed in the Adapter configuration, but when it comes to configuring fields, I receive a 'Read timed out' error. I'm using 'connect-configs' to check if it will work; there is definitely a single JSON object with 3 fields, so there is no chance of a lack of data. It was also checked with a topic with some data from SQL (JDBC connector data).

I've tried all JSON settings, and nothing works for me.

I'm using StreamPipes 0.93.0 and Kafka 3.6.0.

Timeout error message:

![image](https://github.com/apache/streampipes/assets/73185784/4a8cad9f-12dd-4a7c-9099-9e8854825a53)

**Comment by tenthe** [comment]
Hi @mihau1211,
Thank you for your detailed description. Actually, everything you have described looks good to me. I don't see the problem yet, so here are a few more questions:
- If I understand you correctly, the topics are recognized correctly, right?
- While the schema is supposed to be recognized new events are published on the topic? (To recognize the schema, the system needs to receive a new event)
- Do other adapters (e.g. the machine data simulator) work with your instance?
- Can you possibly send an example json to see if there could be a problem.

Cheers,
Philipp

**Reply by mihau1211** [reply]
Hi @tenthe, 
thanks for responding. Answering your questions:
1. Yes, topics are recognized correctly.
2. Firstly, I was just waiting to recognize the schema, now I also checked if publishing new records on the topic would help, but it didn't. Still the same issue.
3. Yes, checked the data simulator and some custom adapters and they were working fine.
4. Yes, here you go:
`{
    "schema": {
        "type": "struct",
        "fields": [
            {
                "type": "int32",
                "optional": false,
                "field": "id"
            },
            {
                "type": "int64",
                "optional": true,
                "name": "org.apache.kafka.connect.data.Timestamp",
                "version": 1,
                "field": "timestamp"
            },
            {
                "type": "string",
                "optional": true,
                "field": "name"
            }
        ],
        "optional": false,
        "name": "test"
    },
    "payload": {
        "id": 1,
        "timestamp": 1701875846033,
        "name": "name1"
    }
}
`

Additionally, I'm using the new topic that is produced by Kafka Connect JDBC Source connector.

EDIT:

In the first post, I didn't notice that it was a different error than usual. Most time I receive that kind of error:

![image](https://github.com/apache/streampipes/assets/73185784/3951fbfd-e6d6-47e8-8030-2ae80f594921)

**Reply by mihau1211** [reply]
Update:

The issue seems to be caused by missing the guess schema endpoint. In the backend container logs, there was a line saying that there was a request for http://172.31.0.6:8090/api/v1/worker/guess/schema, but I couldn't find it on http://172.31.0.6:8090 docs. Accessing the endpoint through the browser returns Method Not Allowed. 

I found also a similar issue on the repository but it was closed without a solution: [316](https://github.com/apache/streampipes/issues/316)

**Reply by tenthe** [reply]
Hi @mihau1211,

first of all, your json event looks good. I tested it with the file stream adapter and it worked.
Maybe a little hint. Your original event contains schema and payload information. In StreamPipes, we mainly focus on events that contain payload key/value pairs in a flat structure.
To transform the event schema, you can drag and drop the nested event properties to the top level and remove the other properties. See the screenshots below.

As for the Kafka error, I need to investigate it further to better understand the problem. It seems that the adapter is not able to read the events from the Kafka theme.

## Original event schema
<img width="1358" alt="original" src="https://github.com/apache/streampipes/assets/5279561/be39895a-40fb-49a5-bb48-93517f067bb2">

## Move properties with drag and drop
<img width="1358" alt="drag" src="https://github.com/apache/streampipes/assets/5279561/f5aaf272-578f-4122-a676-deb8aaeae702">

## Resulting event schema & event
<img width="1358" alt="final" src="https://github.com/apache/streampipes/assets/5279561/164a0846-e0b9-4651-b57a-e98be864f236">

**Comment by dominikriemer** [comment, marked as answer]
Hi @mihau1211 the `Method Not Allowed` message should be ok since the endpoint only support POST requests, and the browser usually performs a GET.

Thanks for the detailed description of the error you get - one thing which could be worth to check is the Kafka configuration. Sometimes, when the metadata discovery works (which seems to be the case since you get the list of topics), but data cannot be retrieved (which happens when the connect module starts to "guess" the schema, this is due to Kafka settings missing an `advertised_listener` which is accessible from inside the Docker network.

Could this be the case? We can try to figure out where you're running Kafka and check that the broker address returned by Kafka is reachable from the StreamPipes container.

Here is some information on that:
https://www.confluent.io/blog/kafka-client-cannot-connect-to-broker-on-aws-on-docker-etc/

Just a guess...if it's another problem, we can investigate further ;-)

**Reply by mihau1211** [reply]
Thanks a lot for your time guys, I only changed the advertised_listener property in Kafka server.properties and it started to work. It was really strange because I was able to connect to the Kafka server from Streampipes containers via curl.
